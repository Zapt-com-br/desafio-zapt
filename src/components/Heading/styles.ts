import styled, { css, DefaultTheme } from 'styled-components';
import media from 'styled-media-query';

import { HeadingProps } from '.';

export const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
  `,

  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xlarge};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xxlarge};
    `}
  `,

  huge: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.huge};
  `,
};

export const Wrapper = styled.h2<HeadingProps>`
  ${({ theme, size, weight }) => css`
    color: ${theme.colors.black};
    font-weight: ${theme.font[weight!]};

    ${!!size && wrapperModifiers[size](theme)}
  `}
`;
