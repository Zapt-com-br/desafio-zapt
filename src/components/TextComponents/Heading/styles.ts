import styled, { css, DefaultTheme } from 'styled-components';
import { HeadingProps } from '.';

type StyledHeadingProps = Pick<HeadingProps, 'size' | 'bold'>;

const headingModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
    line-height: 3.3rem;
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xlarge};
    line-height: 3.6rem;
  `,
};

export const Heading = styled.h1<StyledHeadingProps>`
  ${({ theme, size, bold }) => css`
    font-weight: ${theme.font.weight[bold!]};
    ${!!size && headingModifiers[size](theme)}
  `}
`;
