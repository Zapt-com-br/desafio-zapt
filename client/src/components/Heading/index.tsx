import theme from 'styles/theme'
import * as S from './styles'

export type HeadingProps = {
  color?: string
  size?: keyof typeof theme.font.sizes
  fontWeight?: 300 | 400 | 500
  lineHeight?: number | string
  level?: 1 | 2 | 3 | 4 | 5 | 6
  children: React.ReactNode
}

const Heading = ({
  color = '#000000',
  size = 'normal',
  fontWeight = 500,
  lineHeight = 1.5,
  level = 1,
  children
}: HeadingProps) => (
  <S.Wrapper
    level={level}
    color={color}
    size={size}
    fontWeight={fontWeight}
    lineHeight={lineHeight}
  >
    {children}
  </S.Wrapper>
)

export default Heading
