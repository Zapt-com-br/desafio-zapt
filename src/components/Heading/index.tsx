import * as S from './styles';

export type HeadingProps = {
  children: React.ReactNode;
  size?: 'small' | 'medium' | 'huge';
  weight?: 'light' | 'normal' | 'medium';
};

const Heading = ({
  children,
  size = 'medium',
  weight = 'normal',
}: HeadingProps) => (
  <S.Wrapper size={size} weight={weight}>
    {children}
  </S.Wrapper>
);

export default Heading;
