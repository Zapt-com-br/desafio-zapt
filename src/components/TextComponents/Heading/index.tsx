import React from 'react';
import * as Styled from './styles';

export type HeadingProps = {
  text: string;
  size?: 'small' | 'large';
  bold?: '300' | '400' | '500';
};

const Heading = ({ text, size = 'small', bold = '400' }: HeadingProps) => (
  <Styled.Heading bold={bold} size={size}>
    {text}
  </Styled.Heading>
);

export default Heading;
