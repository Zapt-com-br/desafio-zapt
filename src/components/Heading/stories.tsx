import { Story, Meta } from '@storybook/react/types-6-0';
import Heading, { HeadingProps } from '.';

export default {
  title: 'Heading',
  component: Heading,
  argTypes: {
    children: {
      type: 'string',
    },
    size: {
      type: '',
    },
  },
} as Meta;

export const Default: Story<HeadingProps> = ({ children }) => (
  <>
    <Heading size="huge">Huge - {children}</Heading>
    <Heading>Medium - {children}</Heading>
    <Heading size="small">Small - {children}</Heading>
  </>
);

Default.args = {
  children: 'Heading',
};
