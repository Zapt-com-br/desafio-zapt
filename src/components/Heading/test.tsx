import { screen } from "@testing-library/react";
import { renderWithTheme } from "utils/tests/helpers";

import Heading from ".";

describe("<Heading />", () => {
  it("should render a white heading by default", () => {
    renderWithTheme(<Heading>Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#ffffff",
    });
  });

  it("should render a black heading when color is passed", () => {
    renderWithTheme(<Heading color="black">Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#000000",
    });
  });

  it("should render a gray heading when gray color is passed", () => {
    renderWithTheme(<Heading color="gray">Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#7B7B7B",
    });
  });
});
