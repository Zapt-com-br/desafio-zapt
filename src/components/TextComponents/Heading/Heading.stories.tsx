import React from 'react';

import { Story, Meta } from '@storybook/react/types-6-0';
import Heading, { HeadingProps } from '.';

export default {
  title: 'Text Components/Heading',
  component: Heading,
} as Meta;

export const HeadingDefault: Story<HeadingProps> = args => (
  <Heading {...args} />
);
