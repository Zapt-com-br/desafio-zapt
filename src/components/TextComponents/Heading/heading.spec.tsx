import React from 'react';
import { screen } from '@testing-library/react';
import { renderWithTheme } from 'helpers/renderWithThame';

import Heading from '.';

describe('<Heading />', () => {
  it('should render heading correctly', () => {
    renderWithTheme(<Heading text="heading" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it('should render heading correctly bold value [300]', () => {
    renderWithTheme(<Heading bold="300" text="heading" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toHaveStyleRule('font-weight', '300');
    expect(element).toMatchSnapshot();
  });

  it('should render heading correctly bold value [400]', () => {
    renderWithTheme(<Heading bold="400" text="heading" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toHaveStyleRule('font-weight', '400');
    expect(element).toMatchSnapshot();
  });

  it('should render heading correctly bold value [500]', () => {
    renderWithTheme(<Heading bold="500" text="heading" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toHaveStyleRule('font-weight', '500');
    expect(element).toMatchSnapshot();
  });

  it('should render heading correctly font size and line height value [small]', () => {
    renderWithTheme(<Heading bold="500" text="heading" size="small" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toHaveStyle({
      'font-size': '2.8rem',
      'line-height': '3.3rem',
    });
    expect(element).toMatchSnapshot();
  });
  it('should render heading correctly font size and line height value [large]', () => {
    renderWithTheme(<Heading bold="500" text="heading" size="large" />);

    const element = screen.getByRole('heading', { name: /heading/i });

    expect(element).toHaveStyle({
      'font-size': '4rem',
      'line-height': '3.6rem',
    });
    expect(element).toMatchSnapshot();
  });
});
