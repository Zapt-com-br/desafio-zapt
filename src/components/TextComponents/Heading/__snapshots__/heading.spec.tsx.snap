// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render heading correctly 1`] = `
.c0 {
  font-weight: 400;
  font-size: 2.8rem;
  line-height: 3.3rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;

exports[`<Heading /> should render heading correctly bold value [300] 1`] = `
.c0 {
  font-weight: 300;
  font-size: 2.8rem;
  line-height: 3.3rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;

exports[`<Heading /> should render heading correctly bold value [400] 1`] = `
.c0 {
  font-weight: 400;
  font-size: 2.8rem;
  line-height: 3.3rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;

exports[`<Heading /> should render heading correctly bold value [500] 1`] = `
.c0 {
  font-weight: 500;
  font-size: 2.8rem;
  line-height: 3.3rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;

exports[`<Heading /> should render heading correctly font size and line height value [large] 1`] = `
.c0 {
  font-weight: 500;
  font-size: 4rem;
  line-height: 3.6rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;

exports[`<Heading /> should render heading correctly font size and line height value [small] 1`] = `
.c0 {
  font-weight: 500;
  font-size: 2.8rem;
  line-height: 3.3rem;
}

<h1
  class="c0"
>
  heading
</h1>
`;
